pkg/plugins/policies/meshtrafficpermission/graph/reachable_services_graph.go

```diff
@@ -0,0 +1,161 @@
+package graph
+
+import (
+	"golang.org/x/exp/maps"
+
+	common_api "github.com/kumahq/kuma/api/common/v1alpha1"
+	mesh_proto "github.com/kumahq/kuma/api/mesh/v1alpha1"
+	"github.com/kumahq/kuma/pkg/core"
+	"github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
+	core_model "github.com/kumahq/kuma/pkg/core/resources/model"
+	"github.com/kumahq/kuma/pkg/plugins/policies/core/matchers"
+	core_rules "github.com/kumahq/kuma/pkg/plugins/policies/core/rules"
+	mtp_api "github.com/kumahq/kuma/pkg/plugins/policies/meshtrafficpermission/api/v1alpha1"
+	"github.com/kumahq/kuma/pkg/plugins/runtime/k8s/controllers"
+	"github.com/kumahq/kuma/pkg/xds/context"
+)
+
+var log = core.Log.WithName("rs-graph")
+
+var SupportedTags = map[string]struct{}{
+	controllers.KubeNamespaceTag: {},
+	controllers.KubeServiceTag:   {},
+	controllers.KubePortTag:      {},
+}
+
+type Graph struct {
+	rules map[string]core_rules.Rules
+}
+
+func NewGraph() *Graph {
+	return &Graph{
+		rules: map[string]core_rules.Rules{},
+	}
+}
+
+func (r *Graph) CanReach(fromTags map[string]string, toSvc string) bool {
+	rule := r.rules[toSvc].Compute(core_rules.SubsetFromTags(fromTags))
+	if rule == nil {
+		return false
+	}
+	action := rule.Conf.(mtp_api.Conf).Action
+	return action == mtp_api.Allow || action == mtp_api.AllowWithShadowDeny
+}
+
+func Builder(resources context.Resources) context.ReachableServicesGraph {
+	services := BuildServices(resources.Dataplanes().Items, resources.ExternalServices().Items)
+	mtps := resources.ListOrEmpty(mtp_api.MeshTrafficPermissionType).(*mtp_api.MeshTrafficPermissionResourceList)
+	return BuildGraph(services, mtps.Items)
+}
+
+func BuildServices(
+	dataplanes []*mesh.DataplaneResource,
+	externalServices []*mesh.ExternalServiceResource,
+) map[string]mesh_proto.SingleValueTagSet {
+	services := map[string]mesh_proto.SingleValueTagSet{}
+	for _, dp := range dataplanes {
+		set := dp.Spec.TagSet()
+		for _, svc := range set.Values(mesh_proto.ServiceTag) {
+			if _, ok := services[svc]; ok {
+				continue
+			}
+			services[svc] = map[string]string{}
+			for tag := range SupportedTags {
+				if values := set.Values(tag); len(values) > 0 {
+					services[svc][tag] = values[0]
```

Comment:
Since the `Values()` method sorts the result `values[0]` can be a value from any inbound, right? This means in `services[svc]` you can assemble a not existing inbound, given:
```yaml
inbounds:
  - port: 9090
    tags:
      kuma.io/service: svc-1
      k8s.kuma.io/namespace: aaa
      k8s.kuma.io/service-port: "9090"
  - port: 8080
    tags:
      kuma.io/service: svc-2
      k8s.kuma.io/namespace: bbb
      k8s.kuma.io/service-port: "8080"
```
in `services` you can get:
```yaml
services:
  svc-1:
    k8s.kuma.io/namespace: bbb
    k8s.kuma.io/service-port: "9090"
```
and it's possible there are no policies that're matched for `svc1:bbb:9090`